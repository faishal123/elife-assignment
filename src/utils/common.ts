import { WeatherType } from "@/constants/types/openWeather";

export const mpsToMph = (mps?: number): number => {
	if (!mps) return 0;
	return Math.round(((mps * 2.23693629) / 3.6 + Number.EPSILON) * 100) / 100;
};

export const getLocation = (onSuccess: (data: GeolocationPosition) => void) => {
	if (navigator.geolocation) {
		navigator.geolocation.getCurrentPosition(onSuccess, (error) => {
			switch (error.code) {
				case error.PERMISSION_DENIED:
					alert("Location request has been denied.");
					break;
				case error.POSITION_UNAVAILABLE:
					alert("Unable to get location information.");
					break;
				case error.TIMEOUT:
					alert("Location request timed out");
					break;
				default:
					alert("An unknown error has occurred.");
					break;
			}
		});
	} else {
		alert("This browser doesn't support geolocation.");
	}
};

export const convertTemperature = (tempK?: number) => {
	if (!tempK) return { C: 0, F: 0 };
	const tempC = Math.round(tempK - 273.15);
	const tempF = Math.round(((tempK - 273.15) * 9) / 5 + 32);
	return {
		C: tempC,
		F: tempF,
	};
};

export const getWeatherAndAtmosphere = (weather?: WeatherType[]) => {
	if (!weather) return { weather: [], atmosphere: [] };
	const grouped = weather.reduce<{
		weather: WeatherType[];
		atmosphere: WeatherType[];
	}>(
		(a, c) => {
			if (c.id < 800 && c.id >= 700) {
				return { ...a, atmosphere: [...a.atmosphere, c] };
			}
			return { ...a, weather: [...a.weather, c] };
		},
		{
			weather: [],
			atmosphere: [],
		},
	);
	if (!grouped.weather.length) {
		grouped.weather = grouped.atmosphere;
	}
	return grouped;
};
