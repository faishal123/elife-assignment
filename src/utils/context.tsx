"use client";

import { createContext, ReactNode, useState } from "react";
import { UnsplashResponseType } from "@/constants/types/unsplash";
import {
	CurrentWeatherResponse,
	ReverseGeoCodingResponse,
	ForecastResponse,
} from "@/constants/types/openWeather";

type DefaultContextValueType = {
	forecastSuccess?: boolean;
	weatherSuccess?: boolean;
	citySuccess?: boolean;
	picturesSuccess?: boolean;
	forecast?: ForecastResponse;
	weathers?: CurrentWeatherResponse;
	pictures?: UnsplashResponseType;
	city?: ReverseGeoCodingResponse;
};

export interface ContextDataType extends DefaultContextValueType {
	tempUnit: "C" | "F";
	toggleTempUnit?: () => void;
}

const DataContext = createContext<ContextDataType>({
	tempUnit: "C",
});

const DataProvider = ({
	children,
	defaultValue,
}: {
	children: ReactNode;
	defaultValue: DefaultContextValueType;
}) => {
	const [tempUnit, setTempUnit] = useState<"C" | "F">("C");

	const toggleTempUnit = () => {
		setTempUnit((prev) => (prev === "C" ? "F" : "C"));
	};
	return (
		<DataContext.Provider value={{ ...defaultValue, tempUnit, toggleTempUnit }}>
			{children}
		</DataContext.Provider>
	);
};

export { DataProvider, DataContext };
