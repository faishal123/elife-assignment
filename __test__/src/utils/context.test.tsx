import { render, act } from "@testing-library/react";
import { useContext } from "react";
import { RenderResult } from "@testing-library/react";
import { DataProvider, DataContext, ContextDataType } from "@/utils/context";

let container: RenderResult;

const TestingComponent = () => {
	const { tempUnit, toggleTempUnit } = useContext(DataContext);

	return (
		<div>
			<button data-testid="toggleButton" onClick={toggleTempUnit}>
				toggle
			</button>
			<div data-testid="tempUnitValue">{tempUnit}</div>
		</div>
	);
};

describe("DataProvider", () => {
	it("should render correctly", () => {
		const { container } = render(
			<DataProvider defaultValue={{}}>
				<p>test</p>
			</DataProvider>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with defaultValue", () => {
		const { container } = render(
			<DataProvider
				defaultValue={{
					forecastSuccess: true,
					weatherSuccess: true,
					citySuccess: true,
					picturesSuccess: true,
				}}
			>
				<p>test</p>
			</DataProvider>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should be able to toggle temp unit", async () => {
		act(() => {
			container = render(
				<DataProvider
					defaultValue={{
						forecastSuccess: true,
						weatherSuccess: true,
						citySuccess: true,
						picturesSuccess: true,
					}}
				>
					<TestingComponent />
				</DataProvider>,
			);
		});

		expect(container.queryByTestId("tempUnitValue")).toHaveTextContent("C");
		const button = container.queryByTestId("toggleButton");

		act(() => {
			if (button) {
				button.click();
			}
		});

		expect(container.queryByTestId("tempUnitValue")).toHaveTextContent("F");

		act(() => {
			if (button) {
				button.click();
			}
		});

		expect(container.queryByTestId("tempUnitValue")).toHaveTextContent("C");
	});
});
