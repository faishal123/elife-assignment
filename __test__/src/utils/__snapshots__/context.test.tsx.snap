// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProvider should match snapshot with Context.Consumer 1`] = `
<p>
  test 
  C
</p>
`;

exports[`DataProvider should render correctly 1`] = `
<p>
  test
</p>
`;

exports[`DataProvider should render with defaultValue 1`] = `
<p>
  test
</p>
`;
